강원대_이진호_Step3 기능 구현

treasure-sky opened the pull request:
아직 테스트 코드를 구현하지 못했습니다! 최대한 빨리 작성해서 올리도록 하겠습니다!

### step3 구현 기능

- [x] option 엔티티 구현
- [x] 연관관계 설정
- [x] DTO 생성 및 변경
- [x] option crud 구현
- [x] 기존 API 수정
- [x] option 수량 감소 기능 구현
- [x] 테스트코드 작성

treasure-sky committed:
docs: step3 구현 기능 정리

treasure-sky committed:
refactor: 테이블 생성 방식 변경

- 기존 schema.sql을 사용하여 테이블을 생성하였음.

- 이를 JPA가 자동으로 생성하도록 변경하였음.

- 이에 따라 schema.sql 을 삭제하고 테스트 데이터 삽입 sql은 테스트 패키지로 이동하였음.

treasure-sky committed:
feat: OptionEntity 구현

treasure-sky committed:
feat: 아이템, 옵션 연관관계 설정

treasure-sky committed:
feat: option dto 생성 및 기존 item dto 변경

treasure-sky committed:
feat: option crud 구현

treasure-sky committed:
feat: 기존 item 추가 API 수정

- item 추가 시 option도 포함하도록 수정하였음.

treasure-sky committed:
refactor: 변수명 가독성 향상

treasure-sky committed:
feat: item 생성 시 옵션도 같이 반환하도록 변경

treasure-sky committed:
refactor: dto의 필요없는 필드 제거

treasure-sky committed:
feat: option 삭제 시 해당 item이 최소 1개는 존재하도록 수정

treasure-sky committed:
feat: 옵션 수량을 감소시키는 메서드 추가

treasure-sky committed:
fix: 아이템 생성 API가 옵션도 포함하여 응답하도록 수정

treasure-sky committed:
fix: item 생성 요청 양식 변경에 따른 버그 수정

treasure-sky commented on src/main/java/gift/item/controller/ItemController.java:
```diff
@@ -56,9 +57,9 @@ public ResponseEntity<PageResponseDto<ItemResponseDto>> findAll(
     @PostMapping
-    public ResponseEntity<ItemResponseDto> createItem(
+    public ResponseEntity<ItemDetailResponseDto> createItem(
         @RequestBody @Valid ItemCreateDto itemCreateDto) {
-        ItemResponseDto dto = itemService.createItem(itemCreateDto);
+        ItemDetailResponseDto dto = itemService.createItem(itemCreateDto);
```
상품당 옵션이 1개 이상 존재해야하기 때문에 상품 생성 시에 상품 정보와 옵션을 한 번에 받아오도록 하였습니다.
이에 따라 응답에도 생성된 옵션을 포함시켰습니다.

hjjae2 commented on src/main/java/gift/item/controller/OptionController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.noContent().build();
+    }
+
+
```
사소하지만 불필요한 줄넘김은 제거해주셔도 좋겠습니다. 🙂

hjjae2 commented on src/main/java/gift/item/dto/ItemCreateDto.java:
```diff
@@ -12,7 +15,10 @@ public record ItemCreateDto(
-    String imageUrl
+    String imageUrl,
+
+    @NotEmpty @Valid
```
`@Valid` 어노테이션은 무슨 목적으로 이 위치에 선언이 된건가요-?

treasure-sky replied:
ItemCreateDto에 OptionCreateDto 타입의 필드를 포함하고 있어, OptionCreateDto 내부의 Bean Validator 어노테이션이 적용된 필드를 검증하기 위해 해당 필드에 @Valid를 사용했습니다.

treasure-sky commented on src/main/java/gift/item/dto/OptionCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+public record OptionCreateDto(
+    @NotBlank(message = "옵션명은 필수입니다.")
+    @Size(max = 50, message = "옵션명은 공백포함 최대 50자까지 입력할 수 있습니다.")
+    @OnlyPermittedSymbols
```
상품명의 제약조건을 구현할 때 만들어 둔 커스텀 어노테이션을 재사용하였습니다.

hjjae2 commented on src/main/java/gift/item/dto/OptionCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+
+public record OptionCreateDto(
+    @NotBlank(message = "옵션명은 필수입니다.")
+    @Size(max = 50, message = "옵션명은 공백포함 최대 50자까지 입력할 수 있습니다.")
+    @OnlyPermittedSymbols
+    String name,
+
+    @NotNull(message = "옵션 수량은 필수입니다.")
+    @Min(value = 1, message = "옵션 수량은 최소 1개 이상이어야 합니다.")
+    @Max(value = 99999999, message = "옵션 수량은 1억 개 미만이어야 합니다.")
+    Integer quantity
+) {
```
깔끔하네요. 👍

treasure-sky commented on src/main/java/gift/item/service/OptionService.java:
```diff
@@ -0,0 +1,123 @@
+        Long itemId = oldOptionEntity.getItem().getId();
+        List<OptionEntity> itemOptions = optionRepository.findByItemId(itemId);
+
+        if (itemOptions.size() <= 1) {
+            throw new LastOptionCannotBeDeletedException(itemId);
+        }
```
마지막 옵션의 경우 삭제하지 못하도록 service 단에서 구현했습니다.

treasure-sky commented on src/main/java/gift/item/service/OptionService.java:
```diff
@@ -0,0 +1,123 @@
+        OptionEntity optionEntity = optionRepository.findById(optionId)
+            .orElseThrow(() -> new OptionNotFoundException(optionId));
+
+        if (decrement <= 0) {
+            throw new IllegalArgumentException("감소량(decrement)은 양수여야 합니다.");
+        }
+
+        if (optionEntity.getQuantity() < decrement) {
+            throw new IllegalArgumentException("현재 수량을 초과하는 decrement 입니다.");
+        }
+
+        if (optionEntity.getQuantity() == decrement) {
+            Long itemId = optionEntity.getItem().getId();
+            List<OptionEntity> itemOptions = optionRepository.findByItemId(itemId);
+            if (itemOptions.size() <= 1) {
+                throw new LastOptionCannotBeDeletedException(itemId);
+            }
+        }
```
수량 감소 기능의 경우, 일단 service 단에서만 구현하였습니다.
따라서, 검증을 service 단에 구현하였습니다.

hjjae2 commented on src/main/java/gift/item/OptionEntity.java:
```diff
@@ -0,0 +1,63 @@
+        return item;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public void setQuantity(Integer quantity) {
+        this.quantity = quantity;
+    }
```
[지난번 리뷰](https://github.com/next-step/spring-gift-enhancement/pull/211#discussion_r2213342019)와 유사합니다. 인자로 넘어온 값이 기능 요구 사항에 명시된 조건을 충족함을 보장할 수 있을까요?

treasure-sky replied:
엔티티의 필드를 값 객체로 변경하여 유효성을 보장하도록 변경했습니다!

hjjae2 commented on src/main/java/gift/item/OptionEntity.java:
```diff
@@ -0,0 +1,63 @@
+    public OptionEntity(String name, Integer quantity, ItemEntity item) {
+        this.name = name;
+        this.quantity = quantity;
+        this.item = item;
```
호출처에서 이 생성자는 호출했으나(= option 객체에 item 객체 설정은 됐으나), 
item 객체의 옵션 객체 목록(options)에 option 객체를 추가하는 로직은 호출되지 않은 상황이 있을 수 있겠어요. (예: `OptionService.createOption`) 

때에 따라 이 상황은 버그로 취급될 수도 있습니다. (예: `ItemService.createItem` 에서 `savedItemEntity.getOptions().addAll(optionEntities);` 미호출)

버그로 취급될 수 있는 상황을 방지하기 위해 어떻게 처리를 해볼 수 있을까요?

treasure-sky replied:
cascade 설정을 해두었기 때문에, 연관관계 설정 후에 item을 저장하도록 변경해서 해결하였습니다!

hjjae2 replied:
> cascade 설정을 해두었기 때문에, 연관관계 설정 후에 item을 저장하도록 변경해서 해결하였습니다!

아래와 같은 코드에서 `itemEntity.getOptions().add(newOptionEntity);` 코드를 누락하는 상황에 대해 어떻게 대처할 수 있을지에 대한 질문이었어요. 말씀주신 방법은 제 질문과는 조금 다른 내용인 것 같아요. (제가 잘못 이해했을까요-? 😅 )

```java
        OptionEntity newOptionEntity = new OptionEntity(
            optionCreateDto.name(),
            optionCreateDto.quantity(),
            itemEntity
        );

        itemEntity.getOptions().add(newOptionEntity);
```

코멘트의 의도는, 옵션 엔티티 객체를 생성할 때 상품의 옵션 목록에 옵션을 추가하는 로직(예: `item.getOptions().addAll(this);`)을 함께 처리하면 지금의 코드보다 코드 실수를 방지할 수 있겠다고 생각하고 코멘트를 드렸습니다.

```java
public OptionEntity(String name, Integer quantity, ItemEntity item) {
        this.name = new OptionName(name);
        this.quantity = new Quantity(quantity);
        this.item = item;
        item.getOptions().addAll(this); // <--
}
```

hjjae2 reviewed (changes requested):
안녕하세요, 진호님! 마지막 단계도 고생하셨습니다.

몇 가지 코멘트 남겼는데요. 테스트 코드 작성과 함께 확인 부탁드릴게요. 🙇

treasure-sky committed:
refactor: 불필요한 개행 삭제

treasure-sky committed:
fix: option 생성 시 저장 로직 개선

- 객체간 연관관계 설정 후 item을 저장하도록 변경하였음.

- cascade 설정으로 item, option 모두 저장됨.

treasure-sky committed:
fix: item 저장 시 option이 저장되지 않는 버그 수정

treasure-sky committed:
refactor: option 엔티티 필드를 값 객체로 변경

- 기존에는 엔티티를 생성하거나 수정할 때 파라미터의 유효성을 보장하지 않았음.
- 필드를 값 객체로 변경하고 해당 객체 안에서 검증을 함으로써 유효한 값을 보장하도록 하였음.

treasure-sky committed:
test: option 테스트 추가

treasure-sky committed:
docs: 테스트코드 작성 체크

treasure-sky commented on src/test/java/gift/Item/OptionServieTest.java:
```diff
@@ -0,0 +1,203 @@
+            .isInstanceOf(ItemNotFoundException.class);
+    }
+
+//    @Test
+//    void 옵션_생성_테스트() {
+//        // given
+//        Long itemId = 1L;
+//        OptionCreateDto createDto = new OptionCreateDto("빨간색", 100);
+//        ItemEntity item = new ItemEntity("테스트 상품", 10000, "https://example.com/image.jpg");
+//        OptionEntity savedOption = new OptionEntity("빨간색", 100, item);
+//
+//        given(itemRepository.findById(itemId)).willReturn(Optional.of(item));
+//        given(optionRepository.save(any(OptionEntity.class))).willReturn(savedOption);
+//
+//        // when
+//        OptionResponseDto result = optionService.createOption(itemId, createDto);
+//
+//        // then
+//        assertAll(
+//            () -> assertThat(result.name()).isEqualTo("빨간색"),
+//            () -> assertThat(result.quantity()).isEqualTo(100)
+//        );
+//        then(optionRepository).should().save(any(OptionEntity.class));
+//    }
```
cascade를 사용했기 때문에 OptionEntity가 저장되었는지 알 수 없는 상황입니다. ItemEntity가 저장됨을 테스트하면 될까요?

hjjae2 replied:
네, 지금은 코드의 행동(예: save 1회 실행)을 검증하고 있기에 itenEmtity 가 저장됨을 테스트 해야겠네요. 아마도 repository 가 목 객체이기에, 코드의 행동을 검증하는 형태의 테스트 코드가 작성된 것으로 보여지는데요.

repository 클래스를 Fake 클래스 객체로 만들어 테스트 코드에서 사용하는 방법도 있습니다. 
이런 식으로 작성하면 repository 에 저장된 데이터의 개수(상태)를 검증할 수 있어요.

**예시**

```java
var itemRepository = new FakeItemRepository();
var optionRepository = new FakeOptionRepository();
var optionService = new OptionService(itemRepository, optionRepository);
...

assertThat(optionRepository.findAll().size()).isEqualTo(1); 
```

**참고** 
* https://www.softwaretestingmagazine.com/knowledge/unit-testing-fakes-mocks-and-stubs/#:~:text=to%20use%20them.-,Fake,-Fakes%20are%20objects
* https://medium.com/@CDehning/when-to-use-fakes-instead-of-mocks-c80188b9a3f1

hjjae2 reviewed (approved):
안녕하세요, 진호님! 코멘트 잘 확인해주셨네요, 마지막 단계까지 고생 많으셨습니다. 🙇 

코멘트 드린 부분은 확인 부탁드리며 이번 미션은 여기서 마무리할게요. 

그동안 고생하셨고 마지막 미션까지 화이팅입니다. 💪
